provers/sp1/driver/src/distributed/worker/protocol.rs

```diff
@@ -0,0 +1,112 @@
+use std::fmt::{Display, Formatter};
+
+use raiko_lib::prover::WorkerError;
+use serde::{Deserialize, Serialize};
+use sp1_core::air::PublicValues;
+
+use crate::sp1_specifics::{
+    Challenger, Checkpoint, Commitments, PartialProofs, ShardsPublicValues,
+};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct WorkerEnvelope {
+    version: u64,
+    data: WorkerProtocol,
+}
+
+impl WorkerEnvelope {
+    pub fn data(self) -> Result<WorkerProtocol, WorkerError> {
+        if self.version != include!("../../../worker.version") {
+            return Err(WorkerError::InvalidVersion);
+        }
+
+        Ok(self.data)
+    }
+}
+
+impl From<WorkerProtocol> for WorkerEnvelope {
+    fn from(data: WorkerProtocol) -> Self {
+        WorkerEnvelope {
+            version: include!("../../../worker.version"),
+            data,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum WorkerProtocol {
+    Request(WorkerRequest),
+    Response(WorkerResponse),
+}
+
+impl From<WorkerRequest> for WorkerProtocol {
+    fn from(req: WorkerRequest) -> Self {
+        WorkerProtocol::Request(req)
+    }
+}
+
+impl From<WorkerResponse> for WorkerProtocol {
+    fn from(res: WorkerResponse) -> Self {
+        WorkerProtocol::Response(res)
+    }
+}
+
+impl Display for WorkerProtocol {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            WorkerProtocol::Request(req) => write!(f, "Request({})", req),
+            WorkerProtocol::Response(res) => write!(f, "Response({})", res),
```

Comment:
This could just be a variable capture instead of passing it to the macro:
```suggestion
            WorkerProtocol::Request(req) => write!(f, "Request({req})"),
            WorkerProtocol::Response(res) => write!(f, "Response({res})"),
```

Reply:
Yeah I should definitely use this feature more, I think it didn't exist when I first learned Rust and I've been stuck with this since